### PR TITLE
fix: harden provider settings updates with manager guard

### LINKED ISSUE #2821: [Security] ProviderService.saveProviderSettings() IDOR — any authenticated user can overwrite any provider's settings (claude assist)
## Summary

`ProviderService.saveProviderSettings()` accepts an arbitrary `providerNo` path parameter and writes a `ProviderSettings` object to that provider's record with no authorization check. Any authenticated session can therefore overwrite the settings of any provider in the system.

```java
@POST
@Path("/settings/{providerNo}/save")
public RestResponse<String> saveProviderSettings(ProviderSettings json,
        @PathParam("providerNo") String providerNo) {
    MiscUtils.getLogger().warn(json.toString());           // also logs settings object
    providerManager.updateProviderSettings(getLoggedInInfo(), providerNo, json);
    return RestResponse.successResponse(null);
}
```

This is a classic Insecure Direct Object Reference (IDOR). The only gate is authentication (valid HTTP session); there is no check that the caller *is* the provider named in the path, nor a privilege check for a settings-write role.

## Impact
- Any logged-in user (including patients or non-clinical staff with limited RBAC) can overwrite settings for any provider in the system
- Settings include preferences and configuration that affect how the provider interacts with the EMR
- A secondary issue: `MiscUtils.getLogger().warn(json.toString())` logs the full `ProviderSettings` object at WARN level on every call
- Priority: 2-High

## Location
`src/main/java/.../webserv/rest/ProviderService.java` line 412

## Recommended Fix
Add an authorization check before the `updateProviderSettings` call:

```java
LoggedInInfo info = getLoggedInInfo();
if (!info.getLoggedInProvider().getProviderNo().equals(providerNo)
        && !securityInfoManager.hasPrivilege(info, "_admin", "w", null)) {
    throw new AccessDeniedException("_admin", "w", null);
}
```

This allows a provider to update their own settings, and allows privileged admins to update any provider's settings.

Related: #2798 (missing hasPrivilege across REST services)

### PR DESCRIPTION
**Pull request by makter20**
## Description

The original endpoint-level IDOR fix was independently merged through #3193 while this PR was in review. This branch has been rebased onto current develop and narrowed to the useful follow-up hardening:

- Preserves the existing self-only ProviderService authorization and HTTP 403 response unchanged.
- Adds the same self-only authorization rule inside ProviderManager2.updateProviderSettings as defense in depth, preventing a future non-REST caller from bypassing the endpoint boundary.
- Fails closed when caller identity or the target provider is missing.
- Records a sanitized audit entry after the settings writes, containing only the target provider identifier and no settings values.
- Adds focused service and manager tests for self-edit success, cross-provider denial, missing caller identity, missing target provider, and persistence short-circuiting.

Cross-provider administration remains out of scope and should use a separate explicit admin endpoint if introduced.

## Related Issues

Fixes #2821

## Testing

mvn -q -Dtest=ProviderServiceUnitTest,ProviderManager2UnitTest test

Result: 5 tests, 0 failures, 0 errors, with the normal Checkstyle gate enabled.

## Attribution

The final commit retains Mahmuda Akter as the primary author.

## Checklist

- [x] Commits are DCO signed
- [x] Conventional commit format
- [x] No patient data included
- [x] Focused tests added
- [x] Rebased onto current develop

### CODE REVIEW

File: src/main/java/io/github/carlos_emr/carlos/webserv/rest/ProviderService.java [hunk 1]
```diff
@@ -109,6 +111,8 @@ public class ProviderService extends AbstractServiceImpl {
     @Autowired
     DemographicManager demographicManager;
 
+    @Autowired
+    SecurityInfoManager securityInfoManager;
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/webserv/rest/ProviderService.java]: Not taking this one for this PR. `ProviderService` is an existing CXF/Spring REST resource with a no-arg constructor and existing field-injected collaborators; converting just the newly used collaborator to constructor injection would be an inconsistent partial migration and broader wiring/test churn outside the IDOR fix.

File: src/main/java/io/github/carlos_emr/carlos/webserv/rest/ProviderService.java [hunk 2]
```diff
@@ -410,9 +414,20 @@ public AbstractSearchResponse<ProviderSettings> getProviderSettings() {
...
+        // uses the session-derived provider (never the request path) and fails closed
+        // when the session provider is null, so a missing identity can never authorize.
+        boolean editingSelf = currentProviderNo != null && currentProviderNo.equals(providerNo);
+        if (!editingSelf
+                && !securityInfoManager.hasPrivilege(loggedInInfo, "_admin", SecurityInfoManager.WRITE, null)) {
+            throw new SecurityException("missing required sec object (_admin)");
+        }
 
-        providerManager.updateProviderSettings(getLoggedInInfo(), providerNo, json);
+        providerManager.updateProviderSettings(loggedInInfo, providerNo, json);
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/webserv/rest/ProviderService.java]: Handled by the follow-up refactor: the non-obvious authorization contract now lives in `SecurityInfoManager.assertCanEditProviderSettings`, where it has contract-level JavaDoc for caller identity, target provider, self-edit vs `_admin`/`_admin.userAdmin`, fail-closed behavior, and `SecurityException`. The service and manager now delegate to that single documented gate with local inline context, which avoids duplicating an authorization contract that could drift.

**Comment by Ben-Heerema** [comment]
@makter20 Looks great! Just being a little nit picky there is one more thing I would do, 

`ProviderManager2.updateProviderSettings` writes the audit log before the DAO updates complete. If one of the later persistence operations fails, the audit trail can record `ProviderManager.updateProviderSettings` even though the settings were not successfully saved. Consider moving `LogAction.addLogSynchronous(...)` after the DAO writes, or making the audit action explicitly represent an attempted update rather than a successful one. I'll leave that up to you!

Thanks!

**Comment by makter20** [comment]
@Ben-Heerema I moved addLogSynchronous after the providerExt in the latest commit, so the audit entry is only recorded once all writes succeed (matching the updateProvider pattern right below it).

**Comment by Ben-Heerema** [comment]
Hey @makter20, I reviewed manually and built the branch locally. The core fix is right! Good intuition and work so far :). 

Two things I'd like addressed:

1. The manager-level check has no test pressure on it.
I ran mutation checks against both gates: removing the gate in `ProviderService` fails both `ProviderServiceUnitTest` tests (good — that half is properly pinned), but removing the gate in `ProviderManager2` leaves **all 58 tests passing**. So the defense-in-depth check, the reason `ProviderManager2` is in this PR, could be deleted in a future refactor and CI would stay green. I confirmed with a throwaway test that the guard does work (on denial, `verifyNoInteractions` holds across all four DAOs). A ~30-line `ProviderManager2UnitTest` with `@InjectMocks` would close it, no Spring context needed.

2. The audit-log commactual behaviour.
It claims `addLogSynchronous` "commits its own row and would not roll back with the settings writes"; the method's own JavaDoc says it runs "in the same thread and transaction as it's being called in" it just calls `oscarLogDao.persist()`. Logging after the writes is still fine, it's the stated reasoning that's wrong.

Nice work! The mutation sanity check you did on your own tests is exactly the right instinct. Happy to approve once the manager-level test and the audit comment are sorted.

**Comment by makter20** [comment]
Thanks @Ben-Heerema; both were great catches. I pushed b2b3ed74 addressing the manager-level test coverage and the audit-log reasoning. Appreciate your review.

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@makter20 Your work looks relatively good but it looks like a larger PR from my supervisor #3193 already handles the ticket as it got swept up in our attempt to fix eforms which took a while. That said, I still believe there is useful additions so I am going to take apart your PR a little and take the useful pieces. I hope this is okay, you will still get the credit. Thank you for your help :). If you want to take on any other tickers we are more than happy to have you!

**Comment by Ben-Heerema** [comment]
@coderabbitai review

File: src/test/java/io/github/carlos_emr/carlos/managers/ProviderManager2UnitTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ * @since 2026-07-30
+ * @see ProviderManager2#updateProviderSettings
+ */
+@ExtendWith(MockitoExtension.class)
+@DisplayName("ProviderManager2 Unit Tests")
+@Tag("unit")
+@Tag("fast")
+@Tag("manager")
+class ProviderManager2UnitTest {
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/managers/ProviderManager2UnitTest.java]: Fixed in 55f68e5d. `ProviderManager2UnitTest` now extends `CarlosUnitTestBase` while retaining `MockitoExtension` and its existing test annotations. Verified with `mvn -q -Dtest=ProviderManager2UnitTest test` (passes).

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

File: src/test/java/io/github/carlos_emr/carlos/webserv/rest/ProviderServiceUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            assertThatThrownBy(() -> service.saveProviderSettings(settings, OTHER_PROVIDER_NO))
+                    .isInstanceOf(WebApplicationException.class)
+                    .extracting(exception -> ((WebApplicationException) exception).getResponse().getStatus())
+                    .isEqualTo(Status.FORBIDDEN.getStatusCode());
+
+            verify(mockProviderManager, never()).updateProviderSettings(any(), any(), any());
+        }
+
+        @Test
+        @DisplayName("should return 403 when session provider is null")
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/webserv/rest/ProviderServiceUnitTest.java:159]: Fixed in d05656bb. The null-session test now targets `CALLER_PROVIDER_NO`, isolating the missing-identity fail-closed behavior from the separate cross-provider mismatch case. Verified with `mvn -q -Dtest=ProviderServiceUnitTest test` (passes).

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

File: src/test/java/io/github/carlos_emr/carlos/webserv/rest/ProviderServiceUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        }
+        return info;
+    }
+
+    @Nested
+    @DisplayName("saveProviderSettings authorization")
+    @Tag("update")
+    class SaveProviderSettings {
+
+        @Test
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/webserv/rest/ProviderServiceUnitTest.java:121]: Fixed in 54612584. Added a dedicated missing-target test that calls `saveProviderSettings(settings, (String) null)`, asserts HTTP 403, and verifies `ProviderManager2.updateProviderSettings` is never invoked. The five focused service/manager tests pass.

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@yingbull ready for review

**Comment by makter20** [comment]
@Ben-Heerema, thank you for giving me the chance to contribute to this issue.

**Comment by Ben-Heerema** [comment]
@makter20 of course! Everything helps! You're welcome to take more tickets should you feel like it! :)

**Comment by yingbull** [comment]
All appreciated, thanks @makter20.

**Review by yingbull — APPROVED** [review]
